templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/configuration/AzureDeveloperCliCredential.java

```diff
@@ -0,0 +1,220 @@
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.util.serializer.JacksonAdapter;
+import com.azure.core.util.serializer.SerializerAdapter;
+import com.azure.core.util.serializer.SerializerEncoding;
+import com.azure.identity.CredentialUnavailableException;
+import com.azure.identity.implementation.IdentityClientOptions;
+import com.azure.identity.implementation.util.LoggingUtil;
+import com.azure.identity.implementation.util.ScopeUtil;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import reactor.core.publisher.Mono;
+
+/**
+ * A credential provider that provides token credentials based on
+ * Azure Developer CLI command.
+ */
+@Immutable
+public class AzureDeveloperCliCredential implements TokenCredential {
+
+    private static final ClientLogger LOGGER = new ClientLogger(AzureDeveloperCliCredential.class);
+    private IdentityClientOptions identityClientOptions;
+
+    private static final String WINDOWS_STARTER = "cmd.exe";
+    private static final String LINUX_MAC_STARTER = "/bin/sh";
+    private static final String WINDOWS_SWITCHER = "/c";
+    private static final String LINUX_MAC_SWITCHER = "-c";
+    private static final String DEFAULT_WINDOWS_SYSTEM_ROOT = System.getenv("SystemRoot");
+    private static final String DEFAULT_MAC_LINUX_PATH = "/bin/";
+    private static final String WINDOWS_PROCESS_ERROR_MESSAGE = "'azd' is not recognized";
+    private static final Pattern LINUX_MAC_PROCESS_ERROR_MESSAGE = Pattern.compile("(.*)azd:(.*)not found");
+    private static final Pattern ACCESS_TOKEN_PATTERN = Pattern.compile("\"token\": \"(.*?)(\"|$)");
+    private static final SerializerAdapter SERIALIZER_ADAPTER = JacksonAdapter.createDefaultSerializerAdapter();
+
+    /**
+     * Creates an AzureDeveloperCliCredential with default identity client options.
+     */
+    public AzureDeveloperCliCredential() {
+        identityClientOptions = new IdentityClientOptions();
+    }
+
+    @Override
+    public Mono<AccessToken> getToken(TokenRequestContext request) {
+        return authenticateWithAzureCli(request)
+            .doOnNext(token -> LoggingUtil.logTokenSuccess(LOGGER, request))
+            .doOnError(error -> LoggingUtil.logTokenError(LOGGER, identityClientOptions, request, error));
+    }
+
+    private Mono<AccessToken> authenticateWithAzureCli(TokenRequestContext request) {
+        StringBuilder azCommand = new StringBuilder("azd auth token --output json --scope ");
+
+        var scopes = String.join(" --scope ", request.getScopes());
+
+        try {
+            ScopeUtil.validateScope(scopes);
```

Comment:
Does this work correctly for more than a single scope?

In that case, this string looks like:
`<scope1> --scope <scope2>` which isn't a valid scope?

Reply:
If this needs to be addressed, maybe we should just refactor the appending of scopes as a loop, which then avoids the  ` --scope` as a separator confusion.

Reply:
I see that you're basing off the implementation in `IdentityClient`. There it does:

`String scopes = ScopeUtil.scopesToResource(request.getScopes());`

which actually just validates there can only be 1 single scope.

https://github.com/Azure/azure-sdk-for-java/blob/99ab4a0b6dcf5062899ce90899cb29084ae21f65/sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java#L298

Reply:
That's a very good question. I would need to ask to Identity team :)
The `AzTokenCredential` calls `scopeToResource` like:
![image](https://user-images.githubusercontent.com/24213737/204933802-73bdc774-febf-43b3-8f9d-03c6bbfdc212.png)

And then `scopeToResource` would throw if there is more than one scope:
![image](https://user-images.githubusercontent.com/24213737/204933866-9dfa4f62-4c1b-4cf2-8ed5-d98f60382118.png)


Which makes me think that a request would never have more than one `scope`.  Is thar correct @chlowell ?

If this is true, We might not need the `join` at all and we could just assume there's only one scope all the times